File: Search a 2D Matrix - Leetcode 74/Search a 2D Matrix - Leetcode 74.py
# Brute Force Solution
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for row in matrix:
            if target in row:
                return True
        
        return False
# Time: O(m * n)
# Space: O(1)

# Optimal Solution
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        t = m * n
        l = 0
        r = t - 1

        while l <= r:
            mid = (l + r) // 2
            mid_i = mid // n
            mid_j = mid % n
            mid_num = matrix[mid_i][mid_j]

            if target == mid_num:
                return True
            elif target < mid_num:
                r = mid - 1
            else:
                l = mid + 1

        return False

# Time Complexity: O(log(m * n))
# Space Complexity: O(1)

File: Search a 2D Matrix - Leetcode 74/Search a 2D Matrix - Leetcode 74.java
public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        int n = matrix[0].length;
        int l = 0;
        int r = m * n - 1;

        while (l <= r) {
            int mid = (l + r) / 2;
            int i = mid / n;
            int j = mid % n;
            int midNum = matrix[i][j];

            if (target == midNum) {
                return true;
            } else if (target < midNum) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return false;
    }
}
